### PR TITLE
Implement IdpfPoplar

### PR DESCRIPTION
**Pull request by divergentdave**
This implements IdpfPoplar, as specified in VDAF-03. Key generation, key evaluation, and all necessary codec traits are implemented. The test suite includes tests against a test vector generated by the Sage implementation. The existing `ToyIdpf` is removed, and parts of the existing incompatible `Poplar1` implementation is stubbed out for now.

See previous discussion in #381. (which GitHub auto-closed upon merging Field255)

### CODE REVIEW

File: src/idpf.rs
Comment (cjpatton) [inline src/idpf.rs:251]: Move these lines out of the loop so that we only compute this once.

**Review by cjpatton — COMMENTED** [review]
Not a complete pass yet.

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:76]: BitSlice is an unsized type, only usable as a reference. I could add a `to_owned()` at conversations to make the memory allocation more explicit, or change the argument type to BitVec or BitBox.

Comment (cjpatton) [reply src/idpf.rs:76]: I would change the argument type to BitVec. That way we don't miss an opportunity to avoid a clone, if the opportunity arises!

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:31]: Here are the different cases:

* `gen()`: The number of bits is zero (I realized that the bits argument is redundant with the input's length, so I'll get rid of this)
* `gen()`: The number of beta values and the input bit length don't match
* `gen()`: The input is empty
* `eval()`: The number of bits is zero
* `eval()`: The prefix is empty
* `eval()`: The length of the prefix is greater than the number of bits
* `eval()`: The number of correction words in the public share doesn't match the number of bits

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:56]: I mainly wanted to provide this as an entry point using `std` types only, so that there was a way to construct inputs without dealing with `bitvec` types. I added `clone_with_suffix()` as a helper for the candidate prefixes case.

File: src/idpf.rs
Comment (cjpatton) [inline src/idpf.rs:411]: Hmm, something looks fishy here. Suppose we use `NoCache`. Do we end up evaluating a given prefix more than once?

```
eval():
   for start_level in 0..bits:
      eval_inner(start_level)

eval_inner(start_level):
   for level in start_level..bits:
       eval_next()
```

**Review by cjpatton — COMMENTED** [review]
This looks pretty good. Two high-level things:
1. I'm a bit confused about the structure of `eval()`. It looks to me like we might be wasting some computation. It might help to make it look a bit more like the reference implementation. (See inline comments.)
2. This PR rips out some code form poplar1.rs that we're going to need later. I suggest leaving poplar1.rs alone in this change.

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:411]: eval() only calls eval_inner() once, and then breaks out of its loop, so we're only doing evaluation at each node once.

**Comment by divergentdave** [comment]
For what it's worth, there is one area that I'm aware wastes computation, that I'm leaving as-is for the purposes of this PR. When we evaluate the IDPF at each level, we always compute the field vector representing each node's output, even though we only use this vector from the last computation out of each prefix. I dabbled in optimizations for this, but I wound up either computing something else twice or copying big AES structs around on the stack.

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:487]: Actual evaluation of the IDPF does start with level zero and `prefix[0]`, but this loop is intentionally operating in the reverse order because it's looking for the node at the greatest level where it can get a cache hit. I'll expand on the comments around this.

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:581]: If I reversed the semantic order of the bit indexes and switched from `Lsb0` to `Msb0`, I think that would cancel out and give us the combination of bit endianness and byte endianness we want.

Comment (divergentdave) [reply src/idpf.rs:581]: Wait, that wouldn't work for cases where we have unused bits, because the leading byte/most significant byte would have the wrong bits filled/empty. I think I'll have to ditch using `bitvec` here to avoid the `reverse()` call, and write the shifting and masking myself.

File: src/idpf.rs
Comment (cjpatton) [inline src/idpf.rs:452]: We should probably assert that agg_id is 0 or 1.

File: src/idpf.rs
Comment (cjpatton) [reply src/idpf.rs:470]: Phew, the comments clear this up. Thanks for clarifying!

File: src/idpf.rs
Comment (cjpatton) [reply src/idpf.rs]: (Sorry to do this.) Thanks for figuring out an alternative. The code is far more complex and therefore less clearly correct. In this instance, I would prefer (possibly) slower code that is more efficient. Would you mind reverting to using `reverse()`?

File: src/idpf.rs
Comment (cjpatton) [reply src/idpf.rs:406]: Comments cleared up my confusion, thanks

**Review by cjpatton — COMMENTED** [review]
Just a few minor comments left.

File: benches/cycle_counts.rs
Comment (cjpatton) [reply benches/cycle_counts.rs]: Bump: @divergentdave addressed this in a way I think works, but more eyes on this would be helpful.

**Review by cjpatton — APPROVED** [review]
Looks good to me. Please be sure to squash before merging. Also, there is one outstanding comment that I'd appreciate another pair of eyes on (cc/ @tgeoghegan).

One final note: It occurs to me that we need to improve the coverage of the test vectors: For IDPF, if both aggregators have identical key shares (as they do in the current test), the correction words will cancel out. This prevents us from certain things, such as serialization of the info string input to the PRG.

I'll address this in the spec (https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/132#issuecomment-1386192389) and send a PR with new test vectors and any patches that are needed.

File: src/field.rs
Comment (tgeoghegan) [inline src/field.rs:424]: This `impl` is available anywhere one is working with a concrete type that implements `FieldElement` via this macro. Should we add `ConstantTimeEq` to `trait FieldElement`'s bounds so that it's available when working generically with `impl FieldElement`?

File: src/lib.rs
Comment (tgeoghegan) [inline src/lib.rs:26]: Clients of crate `prio` should be using the eventual `Poplar1` VDAF, not the underlying `idpf` module. Can this be `pub(crate)` or otherwise restricted?

File: src/vdaf.rs
Comment (tgeoghegan) [inline src/vdaf.rs:56]: I think we have a problem if we need to modify generic interfaces exported from `mod vdaf` to accommodate new VDAFs. Our goal is that someone should be able to implement new VDAFs outside of crate `prio` by implementing traits `Vdaf, Aggregator, Collector, Client`. If they don't get to extend this enum, what are they supposed to do? I think what's called for here is something like `codec::CodecError`'s `Other(#[source] Box<dyn Error + 'static + Send + Sync>)` variant.

That being said, this problem already existed (`VdafError::Flp` is arguably specific to the `Prio3` family), so I don't think it should be fixed in this PR: #424.

File: src/idpf.rs
Comment (tgeoghegan) [inline src/idpf.rs:56]: Seems odd that we have `From<BitVec>` and `From<BitBox>` below but these aren't `From<&[u8]>` and `From<&[bool]>`, but I'm OK with either.

File: src/idpf.rs
Comment (tgeoghegan) [inline src/idpf.rs:67]: AFAICT, [`Extend::extend`](https://doc.rust-lang.org/std/iter/trait.Extend.html#tymethod.extend) won't honor the `Vec`'s capacity. Won't this grow `vec` by more than one bit if `suffix.len() > 1`?

File: benches/cycle_counts.rs
Comment (tgeoghegan) [reply benches/cycle_counts.rs]: I might not have done it with macros, but this seems fine to me.

File: src/lib.rs
Comment (cjpatton) [reply src/lib.rs:26]: I think it's useful to export. For example, I might end up implementing (a prototype of) Doplar elsewhere. However since it's behind `experimental` it won't show up with the normal feature set anyway.

Incidentally: What's our goal for API breaking changes for `experimental`? Do we care if we make breaking changes within the same version?

Comment (divergentdave) [reply src/lib.rs:26]: We will also need to either export or wrap `IdpfInput`, for use in the eventual Poplar1 API.

File: src/idpf.rs
Comment (divergentdave) [reply src/idpf.rs:56]: This was done to avoid implicit copies in from/into conversions, see [this thread](https://github.com/divviup/libprio-rs/pull/414#discussion_r1072891268).

File: src/lib.rs
Comment (tgeoghegan) [reply src/lib.rs:26]: > Incidentally: What's our goal for API breaking changes for `experimental`? Do we care if we make breaking changes within the same version?

IMO anything experimental is unstable, i.e. the API can break without a major version change.

> We will also need to either export or wrap `IdpfInput`, for use in the eventual Poplar1 API.

Oh, OK, that's good enough for me. Just wanted to make sure we're deliberately committing to public API surface.

File: benches/cycle_counts.rs
Comment (tgeoghegan) [reply benches/cycle_counts.rs]: To be clear: I'm not arguing against using macros here, it's just that I'm not as good at them as David is, so it's not the tool I would have reached for.

File: src/vdaf.rs
Comment (cjpatton) [reply src/vdaf.rs:56]: @divergentdave seems to be following a design pattern already in this crate: Each module has its own error type. That said I agree. And the error cases are generic enough that I think we could just return `VdafError` instead of creating our own error type.

Comment (tgeoghegan) [reply src/vdaf.rs:56]: I think it's correct for `mod Idpf` to define `IdpfError`, but it's not correct for that to appear as a variant in `VdafError`, which should be generic and not contain any VDAF's specifics.

Comment (cjpatton) [reply src/vdaf.rs:56]: OOoh sorry I misunderstood.

File: src/idpf.rs
Comment (tgeoghegan) [reply src/idpf.rs:67]: The doccomment needs to be updated as clearly this can append more than "one bit".

**Comment by cjpatton** [comment]
@divergentdave can you rebase?

**Review by tgeoghegan — APPROVED** [review]

**Pull request merged**
